apps/passport-server/src/services/issuanceService.ts

```diff
@@ -301,8 +313,128 @@ export class IssuanceService {
     return storedCommitment;
   }
 
-  private async ticketDataToTicketPCD(
+  /**
+   * Fetch all DevconnectPretixTicket entities under a given user's email.
+   */
+  private async issueDevconnectPretixTicketPCDs(
+    request: IssuedPCDsRequest
+  ): Promise<EdDSATicketPCD[]> {
+    return traced(
+      "IssuanceService",
+      "issueDevconnectPretixTicketPCDs",
+      async (span) => {
+        const commitmentRow = await this.checkUserExists(request.userProof);
+        const email = commitmentRow?.email;
+        if (commitmentRow) {
+          span?.setAttribute(
+            "commitment",
+            commitmentRow?.commitment?.toString() ?? ""
+          );
+        }
+        if (email) {
+          span?.setAttribute("email", email);
+        }
+
+        if (commitmentRow == null || email == null) {
+          return [];
+        }
+
+        const commitmentId = commitmentRow.commitment.toString();
+        const ticketsDB = await fetchDevconnectPretixTicketsByEmail(
+          this.context.dbPool,
+          email
+        );
+
+        const tickets = await Promise.all(
+          ticketsDB
+            .map((t) => IssuanceService.ticketRowToTicketData(t, commitmentId))
+            .map((ticketData) => this.getOrGenerateTicket(ticketData))
+        );
+
+        span?.setAttribute("ticket_count", tickets.length);
+
+        return tickets;
+      }
+    );
+  }
+
+  private async getOrGenerateTicket(
     ticketData: ITicketData
+  ): Promise<EdDSATicketPCD> {
+    return traced("IssuanceService", "getOrGenerateTicket", async (span) => {
+      span?.setAttribute("ticket_id", ticketData.ticketId);
+      span?.setAttribute("ticket_email", ticketData.attendeeEmail);
+      span?.setAttribute("ticket_name", ticketData.attendeeName);
+
+      const cachedTicket = await this.getCachedTicket(ticketData);
+
+      if (cachedTicket) {
+        return cachedTicket;
+      }
+
+      logger(`[ISSUANCE] cache miss for ticket id ${ticketData.ticketId}`);
+
+      const generatedTicket = await IssuanceService.ticketDataToTicketPCD(
+        ticketData,
+        this.eddsaPrivateKey
+      );
+
+      try {
+        this.cacheTicket(generatedTicket);
+      } catch (e) {
+        this.rollbarService?.reportError(e);
+        logger(
+          `[ISSUANCE] error caching ticket ${ticketData.ticketId} ` +
+            `${ticketData.attendeeEmail} for ${ticketData.eventId} (${ticketData.eventName})`
+        );
+      }
+
+      return generatedTicket;
+    });
+  }
+
+  private static async getTicketCacheKey(
+    ticketData: ITicketData
+  ): Promise<string> {
+    const ticketCopy: any = { ...ticketData };
+    delete ticketCopy.timestampSigned;
+    const hash = await getHash(JSON.stringify(ticketCopy));
+    return hash;
+  }
```

Comment:
I would naively assume that if we want to cache something, we stringify the thing we want to cache, and then take a cache of that string as the key. Here, we are taking a subset of the PCD (just the `ticket` part of the `claim`), deleting a field, then stringifying the remainder, then hashing that and using that as the key. A comment explaining why `timestampSigned` is problematic for caching would be good, especially if we have to use this pattern again.

Reply:
> A comment explaining why timestampSigned is problematic for caching would be good, especially if we have to use this pattern again.

done

> I wonder if, when caching objects of a given PCD type, we should include the PCD package's version number in the hash.

I think for now what I'm going to do is just clear the cache whenever this happens. We don't have a good way to get the pcd package version from a pcdpackage.